addons/sale/models/sale_order_line.py

```diff
@@ -420,16 +420,20 @@ def _get_custom_compute_tax_cache_key(self):
 
     @api.depends('product_id', 'product_uom', 'product_uom_qty')
     def _compute_pricelist_item_id(self):
-        for line in self:
-            if not line.product_id or line.display_type or not line.order_id.pricelist_id:
-                line.pricelist_item_id = False
-            else:
-                line.pricelist_item_id = line.order_id.pricelist_id._get_product_rule(
-                    line.product_id,
-                    line.product_uom_qty or 1.0,
-                    uom=line.product_uom,
-                    date=line._get_order_date(),
-                )
+        no_item_lines = self.filtered(
+            lambda rec: not rec.product_id or rec.display_type or not rec.order_id.pricelist_id
+        )
+        other_lines = self - no_item_lines
+        no_item_lines.update({"pricelist_item_id": False})
+        for (order, uom, qty), lines in groupby(
+            other_lines,
+            lambda rec: (rec.order_id, rec.product_uom, rec.product_uom_qty)):
+            pricelist = order.pricelist_id
+            date = order.date_order
+            products = self.env['product.product'].browse([line.product_id.id for line in lines])
```

Comment:
Why don't you rely on the default mapped logic here ?
```suggestion
            products = lines.product_id
```

I believe that your current implementation might lead to duplicates in `products`.

Reply:
lines is a list of records, not a recordset (see https://github.com/odoo/odoo/blob/16.0/odoo/tools/misc.py#L1241) but I see your point, and will fix.

Reply:
fixed